### PR TITLE
Revision of the branch

### PR DESCRIPTION
**Pull request by oesteban**
I have revised the utilities and the nipype interface, trying to
simplify the code (without missing functionality).

Some assumptions that can be made on inputs:
* RASB tables will be BIDS-valid, and hence normalized, with absolute
b-vals, etc.
* bvec+bval will be in FSL format (as mandated by BIDS), and hence in image coordinates.

In general, I've tried to remove repetition of code sections, added
doctests that will serve for documentation, minimized dependencies,
checked code style.

I haven't gone through the tests, but they will need a deep revision.
I would recommend using pytest fixtures to minimize the lines of code
and automate some clerical tasks (e.g., setting up data, changing to a
temporal folder, etc.).

### CODE REVIEW

**Comment by oesteban** [comment]
This is just a draft, tomorrow I'll continue with this.

File: dmriprep/conftest.py [hunk 1]
```diff
@@ -0,0 +1,21 @@
+"""py.test configuration."""
```
Comment (oesteban) [inline dmriprep/conftest.py]: This file is just boilerplate for pytest to automate things, could be outside your PR (actually I think I'm going to add this to master before your PR).

File: dmriprep/data/tests/bval [hunk 2]
```diff
@@ -0,0 +1,105 @@
+0.000000000000000000e+00
```
Comment (oesteban) [inline dmriprep/data/tests/bval]: A test bval file, just to have something in place.

File: dmriprep/data/tests/bvec [hunk 3]
```diff
@@ -0,0 +1,3 @@
...63e-01 1.939680948723744580e-01 2.359970500553113526e-01 8.844064781865559333e-01 -4.631901786104944962e-01 -7.000430539717695799e-01 0.000000000000000000e+00 2.000220036306656346e-01 -5.496906112542929712e-01 6.697194463156821431e-01 -2.370055696963325187e-01 -9.603842305537075852e-01 -4.900384695298463239e-02 -7.261448803568887334e-01 6.832152467094113613e-01 0.000000000000000000e+00 -8.447360612148637760e-01 7.303634462442597508e-01 5.099678730359662182e-02 1.799958601428245167e-02 4.441819297579839287e-01 9.895082430076975744e-01 4.698200933497754805e-01 -4.118960093847884307e-01 0.000000000000000000e+00 5.519743337902043301e-01 1.229201892472143037e-01 8.483882264407188734e-01 3.411076364416253948e-01 -3.610025270265337927e-01 4.719905602831905300e-01 8.560779066344169630e-01 -7.969294748620739943e-01 0.000000000000000000e+00 -4.670873535034215651e-01 -1.300449382906106351e-02 -8.824492810552004851e-01 -1.700067153978899667e-02 4.420322695335568985e-01 -3.650675437439669935e-01 -9.770420137099036229e-01 4.061086485971938109e-01 0.000000000000000000e+00 6.271361033032277765e-01 3.539851329366184296e-01 6.582818049316773168e-01 -4.229953470767730539e-01 -2.119635454050159407e-01
```
Comment (oesteban) [inline dmriprep/data/tests/bvec]: An image-coordinates bvec file, consistent with the rasb `dwi.tsv` that is in RAS+ coordinates, using the affine of the fake nifti I also added (`dmriprep/data/tests/dwi.nii.gz`).

File: dmriprep/data/tests/dwi.tsv [hunk 4]
```diff
@@ -0,0 +1,106 @@
+# R	A	S	B
```
Comment (oesteban) [inline dmriprep/data/tests/dwi.tsv]: A rasb file. You can assume this file is kosher because the bids-validator will make all the hard work.

File: dmriprep/interfaces/vectors.py [hunk 5]
```diff
@@ -1,91 +1,101 @@
...
+from ..utils.vectors import DiffusionGradientTable, B0_THRESHOLD, BVEC_NORM_EPSILON
+
+
+class _CheckGradientTableInputSpec(BaseInterfaceInputSpec):
     dwi_file = File(exists=True, mandatory=True)
-    fbval = File(exists=False, mandatory=False)
-    fbvec = File(exists=False, mandatory=False)
-    rasb_tsv_file = File(exists=False, mandatory=False)
-    rescale = traits.Bool(True, usedefault=True)
-    save_fsl_style = traits.Bool(True, usedefault=True)
```
Comment (oesteban) [inline dmriprep/interfaces/vectors.py:14]: You were worried about too many things. Code a class that does just one thing, and you know 100% it does that one thing right.

File: dmriprep/interfaces/vectors.py [hunk 6]
```diff
@@ -1,91 +1,101 @@
...
+
+
+class CheckGradientTable(SimpleInterface):
+    """
+    Ensure the correctness of the gradient table.
+
+    Example
+    -------
+
+    >>> os.chdir(tmpdir)
```
Comment (oesteban) [inline dmriprep/interfaces/vectors.py:38]: run

```
pytest -vv --doctest-modules dmriprep/
```
and you see these doctests running!

File: dmriprep/interfaces/vectors.py [hunk 7]
```diff
@@ -1,91 +1,101 @@
...
+    >>> check.outputs.full_sphere
+    True
+    >>> newrasb = np.loadtxt(check.outputs.out_rasb, skiprows=1)
+    >>> oldrasb = np.loadtxt(str(data_dir / 'dwi.tsv'), skiprows=1)
+    >>> np.allclose(newrasb, oldrasb, rtol=1.e-3)
+    True
+
+    """
+
+    input_spec = _CheckGradientTableInputSpec
```
Comment (oesteban) [inline dmriprep/interfaces/vectors.py:62]: Underscores hide these classes from the name resolvers and you can also leave the docstring unset for them.

File: dmriprep/interfaces/vectors.py [hunk 8]
```diff
@@ -1,91 +1,101 @@
...
-            fbval = self.inputs.fbval
-            fbvec = self.inputs.fbvec
-        else:
-            fbval = None
-            fbvec = None
-
-        if isdefined(self.inputs.rasb_tsv_file):
-            rasb_tsv_file = self.inputs.rasb_tsv_file
-        else:
-            rasb_tsv_file = None
```
Comment (oesteban) [inline dmriprep/interfaces/vectors.py:49]: this was a lot of massaging, very hard to follow when reading.

File: dmriprep/interfaces/vectors.py [hunk 9]
```diff
@@ -1,91 +1,101 @@
...
-        self._results['is_hemi'] = is_hemi
-        self._results['pole'] = pole
-        self._results['slm'] = slm
-        self._results['b0_ixs'] = b0_ixs
-        self._results['fbval_out_file'] = fbval_out_file
-        self._results['fbvec_out_file'] = fbvec_out_file
-        self._results['rasb_tsv_out_file'] = vt.write_rasb_tsv()
+        rasb_file = _undefined(self.inputs, 'in_rasb')
+
+        table = DiffusionGradientTable(
```
Comment (oesteban) [inline dmriprep/interfaces/vectors.py:68]: this is why you want the right data structure in place: it makes it look so easy.

File: dmriprep/interfaces/vectors.py [hunk 10]
```diff
@@ -1,91 +1,101 @@
...
+            bvecs=_undefined(self.inputs, 'in_bvec'),
+            bvals=_undefined(self.inputs, 'in_bval'),
+            rasb_file=rasb_file,
+            b_scale=self.inputs.b_scale,
+            bvec_norm_epsilon=self.inputs.bvec_norm_epsilon,
+            b0_threshold=self.inputs.b0_threshold,
+        )
+        pole = table.pole
+        self._results['pole'] = tuple(pole)
+        self._results['full_sphere'] = np.all(pole == 0.0)
```
Comment (oesteban) [inline dmriprep/interfaces/vectors.py:79]: If the pole (0., 0., 0.) for the full-sphere, why adding checks and checks on it? Just do it once where you really care about it.

File: dmriprep/interfaces/vectors.py [hunk 11]
```diff
@@ -1,91 +1,101 @@
...
+        self._results['b0_ixs'] = np.where(table.b0mask)[0].tolist()
+
+        if rasb_file is None:
```
Comment (oesteban) [inline dmriprep/interfaces/vectors.py:82]: If you convince me about running the checks on rasb files, this section might go away. But I would prefer to first get your PR merged and then revisit this if you manage to make me feel overconfident to fully trust the bids-validator on this one.

File: dmriprep/interfaces/vectors.py [hunk 12]
```diff
@@ -1,91 +1,101 @@
...
+
+        cwd = Path(runtime.cwd).absolute()
+
+        table.to_filename(bvecs=cwd / 'bvec', bvals=cwd / 'bval')
+        self._results['out_bval'] = str(cwd / 'bval')
+        self._results['out_bvec'] = str(cwd / 'bvec')
         return runtime
+
+
+def _undefined(objekt, name, default=None):
```
Comment (oesteban) [inline dmriprep/interfaces/vectors.py:97]: this should be part of nipype probably :P

File: dmriprep/utils/vectors.py [hunk 13]
```diff
@@ -1,482 +1,346 @@
...
+    @property
+    def bvals(self):
+        return self._bvals
+
+    @property
+    def normalized(self):
+        return self._normalized
+
+    @affine.setter
+    def affine(self, value):
```
Comment (oesteban) [inline dmriprep/utils/vectors.py:80]: Python setters/getters are cool

File: dmriprep/utils/vectors.py [hunk 14]
```diff
@@ -1,482 +1,346 @@
...
+
+    @bvecs.setter
+    def bvecs(self, value):
+        if isinstance(value, (str, Path)):
+            value = np.loadtxt(str(value)).T
+        else:
+            value = np.array(value, dtype='float32')
+
+        # Correct any b0's in bvecs misstated as 10's.
+        value[np.any(abs(value) >= 10, axis=1)] = np.zeros(3)
```
Comment (oesteban) [inline dmriprep/utils/vectors.py:103]: you had this setting [1, 0, 0] instead [0, 0, 0] - was that intended?

File: dmriprep/utils/vectors.py [hunk 15]
```diff
@@ -1,482 +1,346 @@
...
-        Array containing the indices of the b0s.
-    """
-    from dipy.core.gradients import gradient_table
-    from dmriprep.utils.vectors import check_corruption
-
-    # Perform initial corruption check, correcting for obvious format/character encoding anomalies.
-    bvecs, bvals = check_corruption(bvecs, bvals, B0_NORM_EPSILON)
-
-    # Creating the gradient table
-    gtab = gradient_table(bvals, bvecs, b0_threshold=B0_NORM_EPSILON)
```
Comment (oesteban) [inline dmriprep/utils/vectors.py:35]: why did you need so much machinery from dipy, if this is just about a couple of numpy arrays?

File: dmriprep/utils/vectors.py [hunk 16]
```diff
@@ -1,482 +1,346 @@
...
         pole = np.mean(vertices, axis=0)
         pole /= np.linalg.norm(pole)
-    else:
-        pole = np.array([0.0, 0.0, 0.0])
-    return is_hemi, pole
-
-
-def set_slm(gtab, is_hemi):
-    """
-    Evaluate eddy compatibility and set second level model that specifies the mathematical form for how the
```
Comment (oesteban) [inline dmriprep/utils/vectors.py:234]: No need for this. We can add a function node right before eddy to check this.

File: dmriprep/utils/vectors.py [hunk 17]
```diff
@@ -1,482 +1,346 @@
...
-        # Read b-values and b-vectors from disk Files can be either ‘.bvals’/’.bvecs’ or ‘.txt’ or ‘.npy’ (containing
-        # arrays stored with the appropriate values).
-
-        if Path(self.fbval).exists() and Path(self.fbvec).exists():
-            self.bvals, self.bvecs = read_bvals_bvecs(self.fbval, self.fbvec)
-        else:
-            raise OSError("Either bval or bvec files not found!")
-
-        # Get b order of magnitude
-        self.bmag = int(np.log10(np.max(self.bvals))) - 1
```
Comment (oesteban) [inline dmriprep/utils/vectors.py:394]: this is the default value of dipy's round_bvals - if you're not going to change that behavior just don't overload it (it's quite risky!)

File: dmriprep/utils/vectors.py [hunk 18]
```diff
@@ -1,482 +1,346 @@
...
-                self.bvecs = self.bvecs.T
-            if not self.bvecs.shape[0] == transform_orientation.shape[0]:
-                raise ValueError("Unrecognized bvec format")
-            output_array = np.zeros_like(self.bvecs)
-            for this_axnum, (axnum, flip) in enumerate(transform_orientation):
-                output_array[this_axnum] = self.bvecs[int(axnum)] * float(flip)
-            self.bvecs = output_array.T
-
-        # Get b order of magnitude
-        self.bmag = int(np.log10(np.max(self.bvals))) - 1
```
Comment (oesteban) [inline dmriprep/utils/vectors.py:423]: twice?

File: dmriprep/utils/vectors.py [hunk 19]
```diff
@@ -1,482 +1,346 @@
...
-        bvals = rescale_bval(bvals.astype('float'), bvecs.astype('float'), B0_NORM_EPSILON)
+    Examples
+    --------
+    >>> bvecs2ras(2.0 * np.eye(3), [(1.0, 0.0, 0.0), (-1.0, 0.0, 0.0)]).tolist()
+    [[1.0, 0.0, 0.0], [-1.0, 0.0, 0.0]]
 
-    # Round bvals by bmag.
-    bval_mag_norm = round_bvals(bvals, bmag=bmag)
+    >>> affine = np.eye(4)
+    >>> affine[0, 0] *= -1.0  # Make it LAS
```
Comment (oesteban) [inline dmriprep/utils/vectors.py:315]: Instead of thinking on all the possible combinations of all possible inputs, just write tests where you know the answer. You'll save lots of time.

**Review by oesteban — COMMENTED** [review]
If you don't count the lines from the conftest, the rasb file and the bvecs/bvals, I've managed to reduce the total number of new lines of your PR by around 100. But also, most of those remaining lines are docstrings with doctests, with do not count as lines of code.

The idea is to solve one problem with the fewest lines possible (and no less). Stand by KISS principles ;).

I'll try to update your tests too. However, I spent most of today with your PR. Can we maybe remove the tests from the PR and add it on a different one next week?

**Comment by dPys** [comment]
Will look through this ASAP, and get back to you! Thanks much,
@dPys

File: dmriprep/conftest.py (hunk 1 above)
Comment (dPys) [reply dmriprep/conftest.py]: I like it. Reading about pytest fixtures now-- I've never used them before.

File: dmriprep/data/tests/bvec (hunk 3 above)
Comment (dPys) [reply dmriprep/data/tests/bvec]: Yeah I do think it's good to have at least one of these living in data

File: dmriprep/interfaces/vectors.py (hunk 6 above)
Comment (dPys) [reply dmriprep/interfaces/vectors.py:38]: Doctests is growing on me.

File: dmriprep/utils/vectors.py (hunk 13 above)
Comment (dPys) [reply dmriprep/utils/vectors.py:80]: Ahh, I wasn't sure before whether you had been thinking of setter/getter decorators, but this looks like a better alternative!

File: dmriprep/utils/vectors.py (hunk 14 above)
Comment (dPys) [reply dmriprep/utils/vectors.py:103]: So sort of. Basically the story behind this is that when I was working with NeuroData using the KKI dataset, we discovered that the B0 vectors were acquired orthogonally (and had some weird labeling such as [1, 0, 0], [0, 0, 1]). Often times we also encountered corrupted cases where there was a '10' in place of the 1, so we hard coded those cases to force to 1. But looking back, since they are actually B0's anyway, I think it's probably perfectly fine to force them to [0, 0, 0] no matter what given that they have no diffusion encoding that becomes relevant anyway. In other words, your correction is perfectly fine and probably more robust across datasets.

**Review by dPys — APPROVED** [review]

File: dmriprep/interfaces/vectors.py (hunk 11 above)
Comment (dPys) [reply dmriprep/interfaces/vectors.py:82]: Sounds like a plan

File: dmriprep/interfaces/vectors.py (hunk 12 above)
Comment (dPys) [reply dmriprep/interfaces/vectors.py:97]: That would be nice!

File: dmriprep/utils/vectors.py (hunk 15 above)
Comment (dPys) [reply dmriprep/utils/vectors.py:35]: Not sure what you're referring to exactly?

File: dmriprep/utils/vectors.py (hunk 16 above)
Comment (dPys) [reply dmriprep/utils/vectors.py:234]: Now that the `pole` output parameter is available on the interface, doing this will be easy. It was drafted as an 'all-in-one' function before since there was not way otherwise to keep the hemisphericity information available to downstream eddy-related nodes in the workflow. I suppose I'm seeing some of the value in nipype interfaces now!

File: dmriprep/utils/vectors.py (hunk 18 above)
Comment (dPys) [reply dmriprep/utils/vectors.py:423]: typo!
